nixos/modules/services/security/kanidm.nix

```diff
@@ -0,0 +1,335 @@
+          description = "Kanidm groups that are allowed to login using PAM.";
+          example = "my_pam_group";
+          type = lib.types.listOf lib.types.str;
+        };
+      };
+      description = ''
+        Configure Kanidm unix daemon.
+        See <link xlink:href="https://github.com/kanidm/kanidm/blob/master/kanidm_book/src/pam_and_nsswitch.md#the-unix-daemon">the documentation</link>
+        and <link xlink:href="https://github.com/kanidm/kanidm/blob/master/examples/unixd">example configuration</link>
+        for possible values.
+      '';
+    };
+  };
+
+  config = lib.mkIf (cfg.enable || cfg.enablePam) {
+    assertions =
+      [
+        {
+          assertion = !cfg.enable || ((cfg.serverSettings.tls_chain or null) == null) || (!lib.isStorePath cfg.serverSettings.tls_chain);
+          message = ''
+            <option>services.kanidm.serverSettings.tls_chain</option> points to
+            a file in the Nix store. You should use a quoted absolute path to
+            prevent this.
+          '';
+        }
+        {
+          assertion = !cfg.enable || ((cfg.serverSettings.tls_key or null) == null) || (!lib.isStorePath cfg.serverSettings.tls_key);
+          message = ''
+            <option>services.kanidm.serverSettings.tls_key</option> points to
+            a file in the Nix store. You should use a quoted absolute path to
+            prevent this.
+          '';
+        }
+        {
+          assertion = !cfg.enablePam || options.services.kanidm.clientSettings.isDefined;
+          message = ''
+            <option>services.kanidm.clientSettings</option> needs to be configured
+            for the PAM daemon to connect to the Kanidm server.
+          '';
+        }
+        {
+          assertion = !cfg.enable || (isNull cfg.ensureDomainName
+            -> cfg.settings.role == "WriteReplica" || cfg.settings.role == "WriteReplicaNoUI");
+          message = ''
+            <option>services.kanidm.ensureDomainName</option> can only be set if this instance
+            is not a ReadOnlyReplica. Otherwise the db would inherit it from
+            the instance it follows.
+          '';
+        }
+      ];
+
+    environment.systemPackages = [ pkgs.kanidm ];
+
+    systemd.services.kanidm = lib.mkIf cfg.enable {
+      description = "kanidm identity management daemon";
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network.target" ];
+      serviceConfig = defaultServiceConfig // {
+        StateDirectory = "kanidm";
+        StateDirectoryMode = "0700";
+        ExecStart = "${pkgs.kanidm}/bin/kanidmd server -c ${serverConfigFile}";
+        ExecPreStart = ''
+          ${pkgs.kanidm}/bin/kanidmd domain_name_change \
+            -c ${serverConfigFile} \
+            -n ${cfg.ensureDomainName}
+        '';
+        User = "kanidm";
+        Group = "kanidm";
+
+        AmbientCapabilities = "CAP_NET_BIND_SERVICE";
+        CapabilityBoundingSet = "CAP_NET_BIND_SERVICE";
```

Comment:
> Could not bind to ldap server address [::1]:636 -> 13, kind: PermissionDenied, message: "Permission denied" }

Not sure why yet.

Reply:
Because ports below 1024 are protected

Reply:
Yeah, but that's why we give the service the `CAP_NET_BIND_SERVICE` capability.

>        CAP_NET_BIND_SERVICE
>              Bind a socket to Internet domain privileged ports (port
>              numbers less than 1024).

https://man7.org/linux/man-pages/man7/capabilities.7.html

Reply:
`PrivateUsers=true` was the culprit here, as it creates a user namespace and applies the CAP_NET_BIND_SERVICE only in that namespace. I disabled it for the service and introduced testing for the open ldap port in 46345ff70ed9e041d9f1c0cbd2edcddece4f5f83.